### PR TITLE
Compare assembly and file versions on minor roll forward

### PR DESCRIPTION
**Pull request by steveharter**
This makes DependencyModel changes so that assemblyVersion and fileVersion can be written to the deps.json file for each runtime asset and adds conflict resolution to hostpolicy.dll that uses that metadata during roll-forward cases.

Fixes https://github.com/dotnet/core-setup/issues/3546

The actual metadata is not added to the .deps.json files yet. That will be done once this PR is in and https://github.com/dotnet/sdk/issues/1847 is resolved.

Changes to conflict resolution are documented at https://github.com/dotnet/core-setup/blob/master/Documentation/design-docs/assembly-conflict-resolution.md.

One conflict test has been commented out until the SDK changes are made (the functionality was verified manually) . Additional tests may be added at that time as well.

### CODE REVIEW

File: src/managed/Microsoft.Extensions.DependencyModel/CollectionExtensions.cs [hunk 1]
```diff
@@ -36,5 +36,22 @@ private static IEnumerable<string> GetAssets(IEnumerable<RuntimeAssetGroup> grou
                 .Where(a => string.Equals(a.Runtime, runtime, StringComparison.Ordinal))
                 .SelectMany(a => a.AssetPaths);
         }
+
+        public static IEnumerable<RuntimeFile> GetDefaultRuntimeFiles(this IEnumerable<RuntimeAssetGroup> self) => GetRuntimeFiles(self, string.Empty);
```
Comment (steveharter) [inline src/managed/Microsoft.Extensions.DependencyModel/CollectionExtensions.cs]: Review note: I'm not happy with this naming. These are mirrors of existing methods like `GetDefaultRuntimeAssets` but return RuntimeFile class instead of a string

File: src/test/Microsoft.Extensions.DependencyModel.Tests/JsonAssetions.cs [hunk 2]
```diff
@@ -31,7 +31,7 @@ public JsonAssetions(JToken token)
             var token = Subject[expected];
             Execute.Assertion
                 .ForCondition(token != null)
-                .FailWith($"Expected {Subject} to have property '" + expected + "'");
+                .FailWith("Expected {0} to have property '{1}'", Subject, expected);
```
Comment (eerhardt) [inline src/test/Microsoft.Extensions.DependencyModel.Tests/JsonAssetions.cs]: Why this change?  How about converting it all to interpolated string usage?

File: src/managed/Microsoft.Extensions.DependencyModel/DependencyContextJsonReader.cs [hunk 3]
```diff
@@ -398,6 +398,46 @@ private List<string> ReadPropertyNames(JsonTextReader reader)
...
+
+                runtimeFiles.Add(new RuntimeFile(path, assemblyVersion, fileVersion));
+            }
+
+            reader.CheckEndObject();
+
+            return runtimeFiles;
+        }
+
         private List<RuntimeTargetEntryStub> ReadTargetLibraryRuntimeTargets(JsonTextReader reader)
```
Comment (eerhardt) [inline src/managed/Microsoft.Extensions.DependencyModel/DependencyContextJsonReader.cs]: I think you need to update `runtimeTargets` as well.

Try building an app that references `System.Data.SqlClient/4.4.0` and you'll see what I mean:

```JSON
"System.Data.SqlClient/4.4.0": {
        "dependencies": {
          "Microsoft.Win32.Registry": "4.4.0",
          "System.Security.Principal.Windows": "4.4.0",
          "System.Text.Encoding.CodePages": "4.4.0",
          "runtime.native.System.Data.SqlClient.sni": "4.4.0"
        },
        "runtime": {
          "lib/netstandard2.0/System.Data.SqlClient.dll": {}
        },
        "runtimeTargets": {
          "runtimes/unix/lib/netstandard2.0/System.Data.SqlClient.dll": {
            "rid": "unix",
            "assetType": "runtime"
          },
          "runtimes/win/lib/netstandard2.0/System.Data.SqlClient.dll": {
            "rid": "win",
            "assetType": "runtime"
          }
        },
        "compile": {
          "ref/netstandard2.0/System.Data.SqlClient.dll": {}
        }
      },
```

File: src/corehost/cli/version.h [hunk 4]
```diff
@@ -0,0 +1,45 @@
+// Copyright (c) .NET Foundation and contributors. All rights reserved.
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+
+#ifndef __VERSION_H__
+#define __VERSION_H__
+
+#include "pal.h"
+#include "utils.h"
+
+struct version_t
```
Comment (eerhardt) [inline src/corehost/cli/version.h]: It would be good to document why we now have 2 "version" types.

Are we sure we can't use the existing fx_ver class?

File: src/managed/Microsoft.Extensions.DependencyModel/RuntimeAssetGroup.cs [hunk 5]
```diff
@@ -22,8 +31,35 @@ public RuntimeAssetGroup(string runtime, IEnumerable<string> assetPaths)
...
+        public IReadOnlyList<RuntimeFile> RuntimeFiles
+        {
+            get
+            {
+                if (_runtimeFiles != null)
+                {
+                    return _runtimeFiles;
+                }
+
+                return _assetPaths.Select(path => new RuntimeFile(path, null, null)).ToArray();
```
Comment (eerhardt) [inline src/managed/Microsoft.Extensions.DependencyModel/RuntimeAssetGroup.cs]: It feels wrong for a property to call `ToArray()` every time it is invoked.

File: src/test/Microsoft.Extensions.DependencyModel.Tests/JsonAssetions.cs (hunk 2 above)
Comment (steveharter) [reply src/test/Microsoft.Extensions.DependencyModel.Tests/JsonAssetions.cs]: Previously, if a test constraint fails, it resulted in an exception due to the way fluentassertions handles args, so it would just show the exception instead of the test assertion info:
```
System.FormatException : Input string was not in a correct format.

System.Text.StringBuilder.FormatError()
System.Text.StringBuilder.AppendFormatHelper(IFormatProvider provider, String format, ParamsArray args)
System.String.FormatHelper(IFormatProvider provider, String format, ParamsArray args)
System.String.Format(String format, Object[] args)
FluentAssertions.Execution.MessageBuilder.FormatArgumentPlaceholders(String failureMessage, Object[] failureArgs) in C:\projects\fluentassertions-vf06b\Src\Core\Execution\MessageBuilder.cs: line 66
FluentAssertions.Execution.MessageBuilder.Build(String message, Object[] messageArgs, String reason, ContextDataItems contextData) in C:\projects\fluentassertions-vf06b\Src\Core\Execution\MessageBuilder.cs: line 42
FluentAssertions.Execution.AssertionScope.FailWith(String message, Object[] args) in C:\projects\fluentassertions-vf06b\Src\Core\Execution\AssertionScope.cs: line 190
Microsoft.Extensions.DependencyModel.Tests.JsonAssetions.HaveProperty(String expected) in D:\git\coresetup15\src\test\Microsoft.Extensions.DependencyModel.Tests\JsonAssetions.cs: line 32
```

File: src/corehost/cli/version.h (hunk 4 above)
Comment (steveharter) [reply src/corehost/cli/version.h]: I added comments to version_t something like this:
```
// This type is made to match semantics of System.Version used by tooling and differs from ver_t:
// -- version_t does not require the third segment
// -- version_t does not support the "pre" label behavior
// -- Different terminology; fx_ver_t(major, minor, patch, build) vs version_t(major, minor, build, revision)
```

File: src/managed/Microsoft.Extensions.DependencyModel/RuntimeAssetGroup.cs (hunk 5 above)
Comment (steveharter) [reply src/managed/Microsoft.Extensions.DependencyModel/RuntimeAssetGroup.cs]: Normally this is not hit; it's for compatibility - it has two modes dependency on the ctor invoked. If the "old" ctor is invoked with string but requests RuntimeFile then this would be hit (and vise-versa if the new ctor invoked).

File: src/managed/Microsoft.Extensions.DependencyModel/DependencyContextJsonReader.cs (hunk 3 above)
Comment (steveharter) [reply src/managed/Microsoft.Extensions.DependencyModel/DependencyContextJsonReader.cs]: Thanks

File: src/managed/Microsoft.Extensions.DependencyModel/DependencyContextExtensions.cs [hunk 6]
```diff
@@ -19,6 +19,15 @@ public static IEnumerable<string> GetDefaultNativeAssets(this DependencyContext
...
+        public static IEnumerable<RuntimeFile> GetDefaultNativeRuntimeFiles(this DependencyContext self)
+        {
+            if (self == null)
+            {
+                throw new ArgumentNullException(nameof(self));
+            }
+            return self.RuntimeLibraries.SelectMany(library => library.GetDefaultNativeRuntimeFiles(self));
+        }
+
         public static IEnumerable<string> GetRuntimeNativeAssets(this DependencyContext self, string runtimeIdentifier)
```
Comment (eerhardt) [inline src/managed/Microsoft.Extensions.DependencyModel/DependencyContextExtensions.cs]: Aren't these breaking changes?

File: src/corehost/cli/deps_format.cpp [hunk 7]
```diff
@@ -228,7 +226,8 @@ bool deps_json_t::process_runtime_targets(const json_value& json, const pal::str
                 if (pal::strcasecmp(type.c_str(), deps_entry_t::s_known_asset_types[i]) == 0)
                 {
                     const auto& rid = file.second.at(_X("rid")).as_string();
-                    assets.libs[package.first].rid_assets[rid].by_type[i].vec.push_back(file.first);
+                    deps_asset_t asset(file.first, empty, empty);
```
Comment (eerhardt) [inline src/corehost/cli/deps_format.cpp]: doesn't this need to be updated for the `runtimeTargets` support?

Comment (steveharter) [reply src/corehost/cli/deps_format.cpp]: Yes, thanks.

**Comment by steveharter** [comment]
@eerhardt @nguerrera I addressed all feedback. Can you take a look please. Thanks

File: src/corehost/cli/deps_format.cpp [hunk 8]
```diff
@@ -36,11 +36,20 @@ pal::string_t deps_json_t::get_optional_path(
     if (iter != properties.end())
     {
         path = iter->second.as_string();
+    }
 
-        if (_X('/') != DIR_SEPARATOR)
-        {
-            replace_char(&path, _X('/'), DIR_SEPARATOR);
-        }
+    return path;
```
Comment (eerhardt) [inline src/corehost/cli/deps_format.cpp]: (nit) it seems like `path` is not a great name for this variable.  How about `value`?
